### PR TITLE
Introduce meson.build

### PR DESCRIPTION
**Pull request by superna9999**
Introduce meson.build to be used with the Meson build system, which handles the dependency and advanced compiler options for when used on various linux distributions.

The libftdi .so for example has different names on different distribution, using meson makes sure we can handle those easily, and easily integrate with other package build systems like Yocto or Buildroot.

### CODE REVIEW

**Comment by konradybcio** [comment]
should we settle on just one?

meson is fine in my eyes, it's pretty common

**Comment by superna9999** [comment]
@konradybcio drop makefile ? probably at some point, but we must be confident meson is ok

**Comment by konradybcio** [comment]
please rebase

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -181,6 +195,43 @@ jobs:
...
+    - name: Meson init for i386
+      if: ${{ matrix.variant == 'i386' }}
+      run: |
+        mkdir build
+        CFLAGS="-m32" LDFLAGS="-m32" meson setup . build
+
+    - name: Meson init with cross compile
+      if: ${{ matrix.variant == 'cross-compile' }}
+      run: |
+        # Generate cross compile file (see https://mesonbuild.com/Cross-compilation.html#cross-compilation)
```
Comment (konradybcio) [inline .github/workflows/ci.yml:208]: surely this can be done with a single `echo` and some multiline magic?

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# SPDX-License-Identifier: GPL-2.0
```
Comment (konradybcio) [inline meson.build:1]: so.. would we get rid of makefile? we don't have the budget of AOSP to support 15 different build systems within the same tree

@lumag?

Comment (lumag) [reply meson.build:1]: I'd say, there should be one building system for all the targets

File: .github/workflows/ci.yml (hunk 1 above)
Comment (lumag) [reply .github/workflows/ci.yml:208]: This is cleaner. than magic, so it is better. What can be done is `cat << EOF .... EOF`, but for me this code is fine too.

File: ci/archlinux.sh [hunk 3]
```diff
@@ -22,6 +22,8 @@ pacman -Sy --noconfirm \
 	systemd-libs \
 	make \
 	pkgconf \
+	meson \
+	ninja \
```
Comment (lumag) [inline ci/archlinux.sh:26]: ninja should come as a dependency for most of the distros, doesn't it? So we do not need to install it explicitly.

File: ci/debian.cross-compile.sh [hunk 4]
```diff
@@ -22,6 +22,9 @@ apt install -y --no-install-recommends \
 	libftdi-dev:${ARCH} \
 	libudev-dev:${ARCH} \
 	libyaml-dev:${ARCH} \
+	pkg-config:${ARCH} \
```
Comment (lumag) [inline ci/debian.cross-compile.sh:25]: Quite no. You are installing an executable package for the target architecture. Quite a hack I'd say.

File: ci/debian.cross-compile.sh [hunk 5]
```diff
@@ -22,6 +22,9 @@ apt install -y --no-install-recommends \
...
+	meson \
+	ninja-build \
```
Comment (lumag) [inline ci/debian.cross-compile.sh:27]: You don't have to repeat this, they should be installed already by `ci/debian.sh`.

File: ci/debian.i386.sh [hunk 6]
```diff
@@ -23,6 +23,8 @@ apt install -y --no-install-recommends \
 	libftdi-dev:i386 \
 	libudev-dev:i386 \
 	libyaml-dev:i386 \
+	meson \
+	ninja-build \
```
Comment (lumag) [inline ci/debian.i386.sh:27]: And here

File: meson.build [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+project('cdba', 'c')
+
+client_srcs = ['cdba.c',
```
Comment (lumag) [inline meson.build:5]: And you have lost all the compiler flags carefully picked up by @krzk

**Review by lumag — CHANGES_REQUESTED** [review]
Generally I'm not sure what is the benefit of switching to meson, other than it being 'newer' than make.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (superna9999) [reply .github/workflows/ci.yml:208]: I wasn’t sure it would work from a GH action yaml but I can try

File: ci/archlinux.sh (hunk 3 above)
Comment (superna9999) [reply ci/archlinux.sh:26]: It doesn’t…

Comment (lumag) [reply ci/archlinux.sh:26]: Huh, it's an issue in arch then. Other distros hopefully have it right.

File: ci/debian.cross-compile.sh (hunk 4 above)
Comment (superna9999) [reply ci/debian.cross-compile.sh:25]: Let me check if there’s not a similar package as gcc

File: meson.build (hunk 7 above)
Comment (superna9999) [reply meson.build:5]: I assume meson carefully select nice flags depending on compiler and source code language, but I can check if it’s necessary to add those flags

File: ci/archlinux.sh (hunk 3 above)
Comment (superna9999) [reply ci/archlinux.sh:26]: I had to install it on all fedora and Debian aswell, pretty sure they aren’t on the base system. I’ll check again but I’m 99.99% sure

File: ci/debian.cross-compile.sh (hunk 4 above)
Comment (lumag) [reply ci/debian.cross-compile.sh:25]: Currently `pkg-config` is `Multi-Arch:same`, which means that is is arch-independent. So you can do it, but I'd still consider for it to be a hack ;-)

File: ci/archlinux.sh (hunk 3 above)
Comment (lumag) [reply ci/archlinux.sh:26]: At least on Debian, `meson` depend on `ninja-build`.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (superna9999) [reply .github/workflows/ci.yml:214]: Well it doesn't work due to how the commands are passed from the yaml file

File: ci/archlinux.sh (hunk 3 above)
Comment (superna9999) [reply ci/archlinux.sh]: Ok it seems to be ok removing them

File: meson.build (hunk 7 above)
Comment (superna9999) [reply meson.build:43]: OK so I added them and it fails on older releases...

https://github.com/linux-msm/cdba/actions/runs/6377506014/job/17306303374

It seems the Makefile doesn't add them for xenial & bionic ?

Comment (lumag) [reply meson.build:45]: Yep. See #34

**Comment by superna9999** [comment]
Well, the biggest point about meson it it's very easy integration with package managers, like:

bitbake:
```
inherit meson
```

debian:
```
dh_auto_configure --buildsystem=meson
```

apkbuild:
```
newapkbuild -m https://github.com/linux-msm/cdba
```

arch:
```
makedepends=(meson)

build() {
  arch-meson $pkgname-$pkgver build
  meson compile -C build
}

check() {
  meson test -C build --print-errorlogs
}

package() {
  meson install -C build --destdir "$pkgdir"
}
```

rpm spec:
```
BuildRequires: meson
```

gentoo:
```
inherit meson
```

buildroot:
```
$(eval $(meson-package))
```

openwrt:
```
# include $(INCLUDE_DIR)/meson.mk
 ```

File: meson.build (hunk 7 above)
Comment (superna9999) [reply meson.build:47]: Solved but the meson version on xenial is so old we can't even test the meson version....

File: meson.build [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+				   '-Wduplicated-branches',
+				   '-Wvla-larger-than=1',
+				   '-Walloc-zero',
+				   '-Wunsafe-loop-optimizations',
+				   '-Wcast-align',
+				   '-Wlogical-op',
+				   '-Wjump-misses-init']
+	elif compiler.get_id() == 'clang'
+		# TODO add clang specific options
+		compiler_cflags = []
```
Comment (lumag) [inline meson.build:39]: `-Wnull-dereference -Wdouble-promotion -Wshadow -Wpointer-arith -Wwrite-strings -Wstrict-overflow=4 -Wundef` ?

**Review by lumag — CHANGES_REQUESTED** [review]
Could you please also delete the Makefile? LGTM otherwise

File: meson.build (hunk 8 above)
Comment (superna9999) [reply meson.build:39]: I moved them in the common base_cflags

**Comment by superna9999** [comment]
@lumag done, dropped all references to makefile and switched CodeQL to autobuild

File: meson.build (hunk 8 above)
Comment (lumag) [reply meson.build]: Ack

**Review by lumag — APPROVED** [review]
Could you please also delete the Makefile?

**Pull request merged**